### PR TITLE
Locate the Python interpreter via the py.exe launcher

### PR DESCRIPTION
**Pull request by pfmoore**
On Windows, if the default strategy for finding a versioned interpreter fails, ask the `py.exe` launcher to locate it for us.

### CODE REVIEW

**Comment by pfmoore** [comment]
Some outstanding questions:

* Does this need tests? It probably should, but the existing tests say "Not testing multiple interpreters on Windows" and I can see why - it's going to be quite hard to test this without mocking a lot of stuff (and possibly reqriting some of the code to make it more testable)
* Should this check *replace* the existing hard-coded list of interpreters? In theory, it's more reliable and the `py.exe` launcher should always be present.
* If we do keep both approaches, should the default list or the `py.exe` check be first?

**Comment by jonparrott** [comment]
Thank you so much for doing this, @pfmoore!

> Does this need tests? It probably should, but the existing tests say "Not testing multiple interpreters on Windows" and I can see why - it's going to be quite hard to test this without mocking a lot of stuff (and possibly reqriting some of the code to make it more testable)

I can handle this - it's likely we'll need to mock some stuff for coverage purposes but have a real test for when on an actual windows platform.

> Should this check replace the existing hard-coded list of interpreters? In theory, it's more reliable and the py.exe launcher should always be present.

I think so, yes.

> If we do keep both approaches, should the default list or the py.exe check be first?

Let's just drop the old method.

File: nox/virtualenv.py [hunk 1]
```diff
@@ -62,6 +64,23 @@ def run(self, args, in_venv=True):
...
 
 
+def locate_via_py(v_maj, v_min):
+    """Find the Python executable using the Windows launcher.
+
+    Code taken from tox (tox/interpreters.py).
+    """
+    ver = "-%s.%s" % (v_maj, v_min)
+    script = "import sys; print(sys.executable)"
+    py_exe = distutils.spawn.find_executable('py')
```
Comment (jonparrott) [inline nox/virtualenv.py]: TIL about `distutils.spawn.find_executable`! Should we use that over [`py.path.local.sysfind`](http://py.readthedocs.io/en/latest/path.html#py._path.local.LocalPath.sysfind) which we use in [command](https://github.com/jonparrott/nox/blob/f87525654265afdb3da086d4116fdfb5867154f6/nox/command.py#L35)?

Comment (pfmoore) [reply nox/virtualenv.py]: Honestly, I doubt it matters (I lifted that code from tox unchanged). I'd be inclined to go for consistency and follow current practice by using `py.path.local.sysfind` in both places. I'll change this.

(BTW, there's also `shutil.which` in Python 3.3+. Who said "one obvious way"? :smile:)

Comment (jonparrott) [reply nox/virtualenv.py]: > (BTW, there's also shutil.which in Python 3.3+. Who said "one obvious way"? 😄)

I'd love to make nox 3+ only. Sigh.

Comment (pfmoore) [reply nox/virtualenv.py]: I'll also make the change to drop the old approach, as noted in your other comment. I can do the tests too, if you want, but if you have a preferred style/approach for the tests and would rather cover that yourself, that's fine too.

Comment (jonparrott) [reply nox/virtualenv.py]: We use pytest so as long as it's not terribly different from what's there go for it. :)

**Comment by pfmoore** [comment]
Looking into the coverage failures...

**Comment by pfmoore** [comment]
There's a remaining coverage failure - line 153 in `nox/virtualenv.py`:

```python
        if self.interpreter:
            cmd.extend(['-p', self._resolved_interpreter])
```

I'm not sure why the changes I've made would have caused this line to no longer be executed. @jonparrott do you have any suggestions?

**Comment by jonparrott** [comment]
> I'm not sure why the changes I've made would have caused this line to no longer be executed. @jonparrott do you have any suggestions?

It's my fault, you can disregard that coverage issue.

**Comment by pfmoore** [comment]
I like that sort of answer ;-) In which case I think this is good to go. Could you check it over for me?

**Comment by jonparrott** [comment]
LGTM- I updated the docstring to mention pep 397. I'm gonna wait for CI to merge (I know it'll fail due to coverage, just want to make sure that's the only failure).

**Comment by jonparrott** [comment]
Thank you so much for doing this @pfmoore! 🍰

**Comment by jonparrott** [comment]
All good. Merging. :)

**Comment by pfmoore** [comment]
Glad to help :smile:

**Comment by dhermes** [comment]
:tada: Yay Windows!

File: nox/virtualenv.py [hunk 2]
```diff
@@ -62,6 +62,24 @@ def run(self, args, in_venv=True):
...
+
+    This is based on :pep:397 which details that executing
+    ``py.exe -{version}`` should execute python with the requested
+    version. We then make the python process print out its full
+    executable path which we use as the location for the version-
+    specific Python interpreter.
+    """
+    script = "import sys; print(sys.executable)"
+    py_exe = py.path.local.sysfind('py')
+    if py_exe:
```
Comment (dhermes) [inline nox/virtualenv.py]: @jonparrott Should this be a [`None`-check][1]?

[1]: http://iruntheinternet.com/lulzdump/images/nunchucks-nuns-chuck-norris-nun-12593488599.jpg

Comment (jonparrott) [reply nox/virtualenv.py]: Possibly. If you want you can send a follow-up PR.

Comment (dhermes) [reply nox/virtualenv.py]: Done #59

**Comment by dhermes** [comment]
BTW I am living life happily in Windows land and I can confirm this PR works great (and `nox` was broken-ish before I `pip install git+https://github.com/jonparrott/nox/`)

**Comment by jonparrott** [comment]
Yay!

**Pull request merged**